src/testing/alerts.rs

```diff
@@ -0,0 +1,295 @@
+use crate::{
+    alerts::{run, Alert, AlertConfig, AlertMessage, ForkProof, ForkingNotification},
+    network::Recipient,
+    nodes::{NodeCount, NodeIndex},
+    rmc::Message as RmcMessage,
+    testing::mock::{Data, Hasher64, KeyBox, PartialMultisignature, Signature},
+    units::{ControlHash, FullUnit, PreUnit, UnitCoord},
+    Index, Indexed, NodeMap, Round, Signable, Signed, UncheckedSigned,
+};
+use futures::{
+    channel::{mpsc, oneshot},
+    FutureExt, StreamExt,
+};
+use futures_timer::Delay;
+use log::debug;
+use std::{collections::HashMap, hash::Hash, time::Duration};
+
+type TestMessage = AlertMessage<Hasher64, Data, Signature, PartialMultisignature>;
+type TestAlert = Alert<Hasher64, Data, Signature>;
+type TestNotification = ForkingNotification<Hasher64, Data, Signature>;
+type TestForkProof = ForkProof<Hasher64, Data, Signature>;
+type TestFullUnit = FullUnit<Hasher64, Data>;
+
+enum Input {
+    Incoming(TestMessage),
+    Alert(TestAlert),
+}
+
+#[derive(Debug, PartialEq, Eq, Hash)]
+enum Output {
+    Outgoing(TestMessage, Recipient),
+    Notification(TestNotification),
+}
+
+struct Scenario {
+    n_members: NodeCount,
+    inputs: Vec<Input>,
+    expected: HashMap<Output, usize>,
+}
```

Comment:
The `Scenario` struct seems redundant. All it does is store inputs and outputs (`n_members` is implied by the `keychains` field of `TestCase`). Can we get rid of `Scenario` and replace `TestCase::run(self, run_as: NodeIndex)` with `TestCase::run_scenario(self, run_as: NodeIndex, inputs: Vec<Input>, mut outputs: HashSet<Output>)`? (the outputs are actually unique in each of your tests)

Reply:
I think this is more readable than building up vectors, especially for bigger tests. To illustrate my point and improve the quality I shall add some shortly. <_<

I also planned to use some matchers in there (which would make the logic much less trivial and allow testing for unexpected output), but I couldn't find a good crate for that...

Reply:
I see your point... However the mechanics of a `TestCase` returning a mutable reference to a `Scenario` is quite confusing as well. Maybe instead we can completely get rid of the `TestCase` struct, and replace its methods with top-level functions? We do not need to keep a vector of keyboxes, we can generate them when they are needed? The lifetimes should not be a problem here as all `Signed`s are immediately downgraded to `UncheckedSigned`s.

Reply:
Now some `Signed`s are actually combined to create a multisignature, so that is no longer an option. I smashed `TestCase` and `Scenario` together, although I had to add one more small struct for some more advanced tests.